gesturelibrary/src/double.c

```diff
@@ -3,31 +3,48 @@
 #include "gestureparams.h"
 #include "utils.h"
 
+#include <stdio.h>
+
 double_tap_t double_tap_d[MAX_TOUCHES];
 void (*on_double_tap)(const double_tap_t*) = 0;
 
 void init_double_tap() {
     for (int i = 0; i < MAX_TOUCHES; i++) {
         double_tap_d[i].state = RECOGNIZER_STATE_NULL;
-        double_tap_d[i].t     = 0;
-        double_tap_d[i].x     = 0;
-        double_tap_d[i].y     = 0;
+        double_tap_d[i].group = i + 1;
+        // double_tap_d[i].num_touches = 0;
+        double_tap_d[i].t = 0;
+        double_tap_d[i].x = 0;
+        double_tap_d[i].y = 0;
     }
     on_double_tap = 0;
 }
 
+// gesture_event_t double_tap = {.type = GESTURE_TYPE_DOUBLE_TAP, .get_data = (void* (*)(void))get_double_tap};
+
 // static touch_event_t* prev_event;
 
-static void update_double_taps(double_tap_t* double_tap, const stroke_t* stroke, const tap_t* tap);
+static void update_double_taps(const tap_t* tap, const touch_event_t* event);
+static void update_data0_fields(int index, int group_index, const touch_event_t* event);
 
 void recognize_double_tap(const touch_event_t* event) {
-    (void)event;  // do we need the touch event
+    // pass in the event and the taps incrementally
+    // if a tap is completed, compare it to all of the double taps in the double data array
+    // and check if for any in_progress ones (if it's null or complete, save that index cuz
+    // we'll need it later) and check if it's close enough; if so, change that double_tap to
+    // complete and update the data fields for it; along the way also compare the timestamps
+    // save the index of any within GROUP_TIME_DIFF (we'll group touches this way); if
+    // nothing is found, update the group and replace the existing complete/null with this
+    // new in_progress double tap
+    // change all in_progress to possible
 
-    const stroke_t* strokes = get_stroke();
-    const tap_t* taps       = get_tap();
+    const tap_t* taps = get_tap();
 
     for (int i = 0; i < MAX_TOUCHES; i++) {
-        update_double_taps(double_tap_d + i, strokes + i, taps + i);
+        if ((taps + i)->t == event->t && (taps + i)->x == event->x &&
+            (taps + i)->y == event->y) {  // make sure we're not looking at an old completed tap
```

Comment:
Instead of searching for a touch event that matches a tap, just use the event->group as the index in your `taps` array.